crates/circuit/src/operations.rs

```diff
@@ -374,6 +382,23 @@ impl Operation for StandardGate {
                 }
                 _ => None,
             },
+            Self::U1Gate => match params[0] {
+                Param::Float(val) => Some(aview2(&gate_matrix::u1_gate(val)).to_owned()),
```

Comment:
I've noticed that `&gate_matrix::u1_gate(val)` is not dereferencing `val`, while the other gate calls do: `&gate_matrix::u2_gate(*phi, *lam)`, `u3_gate(*theta, *phi, *lam)`. Could this be related to the failed unit tests?

Reply:
The difference here is that the match is on `params[0]` instead of `params`. The `val` is already an `f64` and you can't dereference that, while for the other gates the angles have `&f64`. 

I haven't dug into the test failures yet, but I suspect it's more likely I screwed up something in the translation. The qasm3 failure at least is because there are no longer fixed object ids for the gates and I need to update the test to account for that (i.e. use a regex).

Reply:
If this is a blocker for #12598 feel free to just rebase it on main and move ahead without this. I can reindex the gates in this PR without too much trouble.

Reply:
I fixed the test failures in: https://github.com/Qiskit/qiskit/pull/12572/commits/65ff00991a3acd6d3619686a9ced942ebced0662 the optimize 1q gates decomposition failures were because the tests were doing:

`gate.definition.global_phase = val`

and that doesn't do anything to the gate object for a standard gate in rust. TBH, I have no idea what this test is trying to do but to fix it I just used a custom gate class instead of `U1Gate`.

Reply:
So `gate.definition.global_phase = val` is another of the "hidden forbidden moves", like `dag_node.op.params[0]`. Do you think we should document it in the reno too? At a first glance people would probably not expect it to fail.

Reply:
I'm not sure it needs one, the definition was never really supposed to be mutable like this as it's an inherent attribute of a gate, changing it doesn't make much sense (for most things it doesn't get used either). This already wouldn't work with singletons as changing the global phase would affect every instance of the gate. I viewed this as test bug, because I have no idea what that test was trying to accomplish. But I can add a release note out of an abundance of caution if you think it's warranted here.

Reply:
Hmmm I think it would be nice to have it documented somewhere, but I am ok just keeping an eye on it, I don't think people refer to the release notes anywhere and I see how documenting every edge case can be a problem.